### PR TITLE
Enable using zephyr,chosen node to determine flash base address and size for RT685/RT595

### PR DESCRIPTION
**Pull request by danieldegrasse**
Enable using `zephyr, chosen` node to determine the flash base address and flash size for the RT685/RT595 EVK. The `FLASH_BASE_ADDRESS` and `FLASH_SIZE` Kconfigs are set with the following precedence:

`FLASH_BASE_ADDRESS`:
- FlexSPI register property: Used when the flash node's parent has the `nxp,imx-flexspi` compatible. This is the default use case
- node reg property: Used as a fallback. This case will be used when building targeting another memory region, such as SRAM

`FLASH_SIZE`:
- flash node's `size` property: Used when the `size` property is present on the chosen flash node, such as when the flash node is the external flash chip on the board. The flash device's size is expressed in bits per the requirement of the `jedec,jesd216` compatible, so the size must first be converted to bytes
- node reg size property: Used as a fallback. This case will be used when building targeting another memory region, such as SRAM

In order to facilitate the required Kconfig changes, several new Kconfig functions have been added. The "size" Kconfig functions have been extended to enable converting a value in bits to kilobytes, megabytes, or gigabytes. Additionally, two new Kconfig functions have been added:
- `dt_node_parent` function: given a node path, returns the path of the node's parent
- `dt_node_has_compat` function: given a node path and a compatible, checks if  the compatible is present on that node

### CODE REVIEW

**Comment by danieldegrasse** [comment]
@galak This PR is intended to supersede #44762, and leaves all `reg` properties untouched. Would love your feedback on this approach before we proceed with it.

**Review by DerekSnell — DISMISSED** [review]

File: dts/bindings/mtd/nxp,imx-flexspi-mx25um51345g.yaml [hunk 1]
```diff
@@ -6,3 +6,14 @@ description: NXP FlexSPI MX25UM51345G
 compatible: "nxp,imx-flexspi-mx25um51345g"
 
 include: ["nxp,imx-flexspi-device.yaml", soc-nv-flash.yaml]
+
+properties:
+  device-base:
+        type: int
+        required: false
+        description: |
+          Base memory mapped address of this flash device. Typically the same
```
Comment (galak) [inline dts/bindings/mtd/nxp,imx-flexspi-mx25um51345g.yaml:15]: Do you mind providing an example (not in the PR, but in this thread).  What would the dts look like for this case?

File: boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts [hunk 2]
```diff
@@ -309,12 +309,18 @@ i2s1: &flexcomm3 {
 	status = "okay";
 	mx25um51345g: mx25um51345g@2 {
 		compatible = "nxp,imx-flexspi-mx25um51345g";
-		size = <DT_SIZE_M(64)>;
+		size = <DT_SIZE_M(512)>;
```
Comment (galak) [inline boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:312]: How about making `size` a const in the `nxp,imx-flexspi-mx25um51345g.yaml`.  Also, add a comment that this is in `bits`.  (personally I prefer to write this as `size = <DT_SIZE_M(64*8)>; /*size in bits */`

**Review by galak — DISMISSED** [review]
In general looks good.  Do want to better understand `device-base` and if `stacked-memories` https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/spi/spi-peripheral-props.yaml#n81 might be useful or not.

**Comment by danieldegrasse** [comment]
@galak I've dropped `device-base` from this PR. The core use case of this property was to support multiple flash chips on one FlexSPI bus, but as the required support for multiple flash chips isn't present in the NXP flexSPI flash driver, I'll leave this support for a later PR.

File: dts/bindings/mtd/nxp,imx-flexspi-mx25um51345g.yaml (hunk 1 above)
Comment (danieldegrasse) [reply dts/bindings/mtd/nxp,imx-flexspi-mx25um51345g.yaml]: I've dropped this property in the latest push, see my comment below

File: soc/arm/nxp_imx/rt5xx/Kconfig.defconfig.series [hunk 3]
```diff
@@ -17,13 +17,29 @@ config NUM_IRQS
...
 config FLASH_BASE_ADDRESS
-	default $(dt_node_reg_addr_hex,/soc/spi@134000,1)
+	default $(dt_node_reg_addr_hex,$(DT_CHOSEN_FLASH_PARENT),1) \
+		if $(dt_node_has_compat,$(DT_CHOSEN_FLASH_PARENT),$(DT_COMPAT_FLEXSPI))
+	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_FLASH))
+
+# Boards must add another default to this Kconfig to indicate the size of their
+# external flash, since the RT5xx has no internal flash.
+config FLASH_SIZE
+	default $(dt_chosen_reg_size_int,$(DT_CHOSEN_Z_FLASH),0,K)
```
Comment (mmahadevan108) [inline soc/arm/nxp_imx/rt5xx/Kconfig.defconfig.series:42]: Why do we need set this board level. Can we get this from the zephyr,flash binding

Comment (danieldegrasse) [reply soc/arm/nxp_imx/rt5xx/Kconfig.defconfig.series:42]: Sure, we can move the FLASH_SIZE setting to SOC level. I'll update this PR.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
+1 to the kconfigfunctions.py changes, but you need to keep doc/build/kconfig/preprocessor-functions.rst in sync

**Comment by danieldegrasse** [comment]
> +1 to the kconfigfunctions.py changes, but you need to keep doc/build/kconfig/preprocessor-functions.rst in sync

Thanks for the reminder, I've updated it

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]

**Comment by carlescufi** [comment]
@danieldegrasse please rebase

**Comment by danieldegrasse** [comment]
> @danieldegrasse please rebase

Done, thanks

**Review by nordicjm — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Pull request merged**